### DISCUSSION TITLE
Allow non-member bitwise operators

### DISCUSSION THREAD

**Discussion by KendallHarter** [hsutter/cppfront#768] [Suggestions]
Currently, bitwise operators are required to be member functions (see [here](https://github.com/hsutter/cppfront/blob/main/source/sema.h#L1295)) and the generated code results in member functions as well.  This is contrary to C++ Core Guideline [C.161](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c161-use-non-member-functions-for-symmetric-operators) which recommends using non-member functions.  I've looked around but can't find any motivation for this restriction and it creates situations where unfixable asymmetry exists:
```
test: type = {
    operator &: (this, _ : i32) -> bool = true;
}

main: () = {
    // Compile error; no way to define in cpp2
    // [[assert: 1 & test()]]
    [[assert: test() & 1]]
}
```

**Comment by hsutter** [comment]
Hmm, I can't remember now why I did that. I've removed that restriction in 37cf0bf1b9320524a6ce8c6eb92d6ed7f135d9be -- thanks!

For comparisons the reason I restricted them to members is because with C++20 (Cpp2-derived) symmetry there's no longer a need to write them as nonmembers. But even there I suppose there could be a motivation to write them as nonmembers in order to get comparisons on a type that doesn't provide them itself -- I'll leave that alone for now but we'll see if it's often needed.

**Comment by hsutter** [comment]
This code now works:

```
test: type = { }

operator &: (_:test, _:i32) -> bool = true;
operator &: (_:i32, _:test) -> bool = true;

main: () = {
    [[assert: 1 & test()]]
    [[assert: test() & 1]]
}
```

Thanks again.
